storage/swift_storage.py

```diff
@@ -1,253 +1,254 @@
 
-        directories, files = self._list_container_objects(container_name, prefix)
+            auth, _ = self._parsed_storage_uri.netloc.split("@")
+            user, key = auth.split(":", 1)
 
-        for directory in directories:
-            if directory.name == prefix:
-                continue
+            if user == "":
+                raise SwiftStorageError(f"Missing username")
 
-            directory_path = directory.name.split('/', 1).pop()
-            target_directory = os.path.join(destination_directory, directory_path)
+            if key == "":
+                raise SwiftStorageError(f"Missing API key")
 
-            if not os.path.exists(target_directory):
-                os.makedirs(target_directory)
+            auth = v2.Password(
+                auth_url=auth_endpoint, username=user, password=key, tenant_name=self.tenant_id)
+
+            keystone_session = session.Session(auth=auth)
+
+            connection = swiftclient.client.Connection(
+                session=keystone_session, os_options=os_options, timeout=DEFAULT_SWIFT_TIMEOUT)
+            self._connection = connection
+        return self._connection
+
+    def get_container_and_object_names(self) -> Tuple[str, str]:
+        _, container = self._parsed_storage_uri.netloc.split("@")
+        object_name = self._parsed_storage_uri.path[1:]
+        return container, object_name
 
-        for file in files:
-            directory = os.path.dirname(file.name.replace(prefix, destination_directory, 1))
+    def _download_object_to_file(
+            self, container: str, object_name: str, out_file: BinaryIO) -> None:
+        connection = self.get_connection()
 
-            if not os.path.exists(directory):
-                os.makedirs(directory)
+        def get_object() -> None:
+            out_file.seek(0)
+            resp_headers, object_contents = connection.get_object(
+                container, object_name, resp_chunk_size=_LARGE_CHUNK)
+            # may need to truncate file if retrying...
+            for object_content in object_contents:
+                out_file.write(object_content)
 
-            retry.attempt(
-                self._cloudfiles.download_object,
-                container_name,
-                file,
-                directory,
-                structure=False)
+        retry_swift_operation(
+            f"Failed to retrieve Swift object {object_name} from container {container}",
+            get_object)
 
-    def _upload_file(self, file_or_path, object_path=None):
-        self._authenticate()
-        container_name, object_name = self._get_container_and_object_names()
-        kwargs = {}
-        mimetype = mimetypes.guess_type(object_name)[0]
-        if mimetype is not None:
-            kwargs["content_type"] = mimetype
+    def _download_object_to_filename(self, container: str, object_name: str, filename: str) -> None:
+        dir_name = os.path.dirname(filename)
+        os.makedirs(dir_name, exist_ok=True)
```

Comment:
It looks like, previously, we were erroring if the directory didn't exist. Is it OK to change this behavior?

Reply:
I am for whatever is parallel to all other storage implementations, since it's supposed to be transparent to the end user regardless of the storage protocol.